### PR TITLE
StorageCluster spec change and updating Noobaa reconcile

### PR DESCRIPTION
**Pull request by vbnrh**
This commit adds changes to StorageCluster api for supporting additional parameters for Noobaa to use external Postgres server and the parameters are passed to Noobaa CR

https://issues.redhat.com/browse/RHSTOR-4749

### CODE REVIEW

File: api/v1/storagecluster_types.go [hunk 1]
```diff
@@ -377,6 +377,31 @@ type MultiCloudGatewaySpec struct {
...
+	PGSecretNamespace string `json:"pgSecretNamespace,omitempty"`
+	// AllowSelfSignedCerts will allow the Postgres server to use self signed certificates to authenticate
+	// +optional
+	AllowSelfSignedCerts bool `json:"allowSelfSignedCerts,omitempty"`
+	// TLSSecret stores the secret name which contains the client side certificates if enabled
+	// +optional
+	TLSSecret string `json:"tlsSecret,omitempty"`
+	// TLSecretNamespace is the namespace where the TLS secret lives
+	// +optional
+	TLSSecretNamespace string `json:"tlsSecretNamespace,omitempty"`
```
Comment (Madhu-1) [inline api/v1/storagecluster_types.go:404]: The namespace mentioned in the spec is the same namespace where storagecluster CR is created or it can be anything?

File: api/v1/storagecluster_types.go [hunk 2]
```diff
@@ -377,6 +377,31 @@ type MultiCloudGatewaySpec struct {
...
+	ExternalPgConfig ExternalPGSpec `json:"externalPgConfig,omitempty"`
+}
+
+type ExternalPGSpec struct {
+	// Enables Noobaa to use external Postgres
+	// +optional
+	Enable bool `json:"enable,omitempty"`
+	// PGSecret stores the secret name which contains connection string of the Postgres server
+	// +optional
+	PGSecret string `json:"pgSecret,omitempty"`
```
Comment (Madhu-1) [inline api/v1/storagecluster_types.go:392]: `PGSecret` to `PGSecretName`

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 3]
```diff
@@ -172,7 +172,7 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 	}
 
 	// Override with MCG options specified in the storage cluster spec
-	if sc.Spec.MultiCloudGateway != nil {
+	if sc.Spec.MultiCloudGateway.DbStorageClassName != "" {
```
Comment (Madhu-1) [inline controllers/storagecluster/noobaa_system_reconciler.go:175]: you still need `nil` check as its a pointer

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 4]
```diff
@@ -193,6 +193,14 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 		nb.Spec.DisableLoadBalancerService = sc.Spec.MultiCloudGateway.DisableLoadBalancerService
 	}
 
+	if sc.Spec.MultiCloudGateway.ExternalPgConfig.Enable {
```
Comment (Madhu-1) [inline controllers/storagecluster/noobaa_system_reconciler.go:196]: you need `nil` check for `sc.Spec.MultiCloudGateway` here as well

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (vbnrh) [reply api/v1/storagecluster_types.go:404]: It can be anything. Noobaa takes in a secret reference so name and namespace can be anything.

File: controllers/storagecluster/noobaa_system_reconciler.go (hunk 4 above)
Comment (vbnrh) [reply controllers/storagecluster/noobaa_system_reconciler.go]: done

File: controllers/storagecluster/noobaa_system_reconciler.go (hunk 3 above)
Comment (vbnrh) [reply controllers/storagecluster/noobaa_system_reconciler.go]: done

File: api/v1/storagecluster_types.go (hunk 2 above)
Comment (vbnrh) [reply api/v1/storagecluster_types.go]: done and same with TLSSecret

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 5]
```diff
@@ -172,7 +172,7 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 	}
 
 	// Override with MCG options specified in the storage cluster spec
-	if sc.Spec.MultiCloudGateway != nil {
+	if sc.Spec.MultiCloudGateway != nil && sc.Spec.MultiCloudGateway.DbStorageClassName != "" {
```
Comment (umangachapagain) [inline controllers/storagecluster/noobaa_system_reconciler.go:175]: Is this change needed for this feature to work?

**Comment by vbnrh** [comment]
Screenshot 2023-11-14 at 7 45 51 PM
Not sure why it is CI is failing. All changes have already been commited

File: api/v1/storagecluster_types.go [hunk 6]
```diff
@@ -377,6 +377,31 @@ type MultiCloudGatewaySpec struct {
...
+
+	// Allows Noobaa to connect to an external Postgres server
+	// +optional
+	ExternalPgConfig *ExternalPGSpec `json:"externalPgConfig,omitempty"`
+}
+
+type ExternalPGSpec struct {
+	// Enables Noobaa to use external Postgres
+	// +optional
+	Enable bool `json:"enable,omitempty"`
```
Comment (umangachapagain) [inline api/v1/storagecluster_types.go:389]: Can we drop this field and assume that we need external postgres when `PGSecretName` and `PGSecretNamespace` are not nil?
I don't see why I would set `enable:false` and provide connection info or vice-versa.

File: controllers/storagecluster/noobaa_system_reconciler.go (hunk 5 above)
Comment (vbnrh) [reply controllers/storagecluster/noobaa_system_reconciler.go:175]: It is not related to the feature but to avoid the loop since it will not be nil if pgsecrets are not nil

File: api/v1/storagecluster_types.go (hunk 6 above)
Comment (vbnrh) [reply api/v1/storagecluster_types.go]: updated. i added the enable to have easy to read checks in future.

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (Madhu-1) [reply api/v1/storagecluster_types.go:401]: i dont know if we need to support this one or not, i assume we always want the users to make use of the namespace where storagecluster is created /odf is installed.

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 7]
```diff
@@ -193,6 +193,14 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 		nb.Spec.DisableLoadBalancerService = sc.Spec.MultiCloudGateway.DisableLoadBalancerService
 	}
 
+	if sc.Spec.MultiCloudGateway != nil && sc.Spec.MultiCloudGateway.ExternalPgConfig != nil && sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretName != "" {
+		nb.Spec.ExternalPgSSLRequired = true
+		nb.Spec.ExternalPgSecret = &corev1.SecretReference{Name: sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretName, Namespace: sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretNamespace}
+		nb.Spec.ExternalPgSSLUnauthorized = sc.Spec.MultiCloudGateway.ExternalPgConfig.AllowSelfSignedCerts
+		if sc.Spec.MultiCloudGateway.ExternalPgConfig.TLSSecretName != "" && sc.Spec.MultiCloudGateway.ExternalPgConfig.TLSSecretNamespace != "" {
+			nb.Spec.ExternalPgSSLSecret = &corev1.SecretReference{Name: sc.Spec.MultiCloudGateway.ExternalPgConfig.TLSSecretName, Namespace: sc.Spec.MultiCloudGateway.ExternalPgConfig.TLSSecretNamespace}
```
Comment (Madhu-1) [inline controllers/storagecluster/noobaa_system_reconciler.go:201]: Do we need a check here to ensure that both name and namespace are set not either one of them?

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 8]
```diff
@@ -193,6 +193,14 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 		nb.Spec.DisableLoadBalancerService = sc.Spec.MultiCloudGateway.DisableLoadBalancerService
 	}
 
+	if sc.Spec.MultiCloudGateway != nil && sc.Spec.MultiCloudGateway.ExternalPgConfig != nil && sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretName != "" {
+		nb.Spec.ExternalPgSSLRequired = true
+		nb.Spec.ExternalPgSecret = &corev1.SecretReference{Name: sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretName, Namespace: sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretNamespace}
+		nb.Spec.ExternalPgSSLUnauthorized = sc.Spec.MultiCloudGateway.ExternalPgConfig.AllowSelfSignedCerts
+		if sc.Spec.MultiCloudGateway.ExternalPgConfig.TLSSecretName != "" && sc.Spec.MultiCloudGateway.ExternalPgConfig.TLSSecretNamespace != "" {
```
Comment (Madhu-1) [inline controllers/storagecluster/noobaa_system_reconciler.go:200]: what if someone sets the TLSSecret but forgets to set PGSecret details, should we error out in that case?

**Review by Madhu-1 — COMMENTED** [review]
Left few questions,LGTM

File: controllers/storagecluster/noobaa_system_reconciler.go (hunk 7 above)
Comment (vbnrh) [reply controllers/storagecluster/noobaa_system_reconciler.go:201]: It will work with either one but just in case added both to for secret ref.

File: controllers/storagecluster/noobaa_system_reconciler.go (hunk 8 above)
Comment (vbnrh) [reply controllers/storagecluster/noobaa_system_reconciler.go:200]: I have changed the check above to see if PGSecret is empty or not. It won't enter this block if it is not set.

File: controllers/storagecluster/noobaa_system_reconciler.go [hunk 9]
```diff
@@ -193,6 +193,14 @@ func (r *StorageClusterReconciler) setNooBaaDesiredState(nb *nbv1.NooBaa, sc *oc
 		nb.Spec.DisableLoadBalancerService = sc.Spec.MultiCloudGateway.DisableLoadBalancerService
 	}
 
+	if sc.Spec.MultiCloudGateway != nil && sc.Spec.MultiCloudGateway.ExternalPgConfig != nil && sc.Spec.MultiCloudGateway.ExternalPgConfig.PGSecretName != "" {
```
Comment (umangachapagain) [inline controllers/storagecluster/noobaa_system_reconciler.go:196]: need to check `PGSecretNamespace!=""` as well.

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (jackyalbo) [reply api/v1/storagecluster_types.go:401]: let's remove the namspace I will do the same for NooBaa repo as well. We can't mount secrets from different namespaces anyway.

Comment (jackyalbo) [reply api/v1/storagecluster_types.go:401]: Also PGSecretNamespace should be removed

File: controllers/storagecluster/noobaa_system_reconciler.go (hunk 9 above)
Comment (vbnrh) [reply controllers/storagecluster/noobaa_system_reconciler.go]: done

**Comment by vbnrh** [comment]
Screenshot 2023-11-16 at 2 21 42 PM

**Review by umangachapagain — APPROVED** [review]

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (Madhu-1) [reply api/v1/storagecluster_types.go:401]: @vbnrh will you be deleting these values from the types?

Comment (vbnrh) [reply api/v1/storagecluster_types.go:401]: @Madhu-1 @jackyalbo I had failed to see these reviews. I will update the values

**Comment by malayparida2000** [comment]
Nooba uses a slightly older version of rook api 
https://github.com/noobaa/noobaa-operator/blob/2b58b1e61bfff592718abede1824b1efdd6c098c/go.mod#L34
In that version rook had only 1 go.mod in it's repo. Now with the latest rook changes rook has separated their repo a bit, The repo has a go.mod & their pkg dir has another go.mod. So when we are trying to bring in this noobaa version it tries to bring in that older rook version which had only 1 go.mod, But now ocs-operator has both the paths for rook added which creates an ambiguous import.

Just add
github.com/rook/rook v1.12.3-0.20230823002913-59e66edf92f7
to the exclude section in the go.mod with a suitable comment and it will work.

**Review by malayparida2000 — APPROVED** [review]
/lgtm
